### DISCUSSION TITLE
GNU compiler issue - IEEE_DIVIDE_BY_ZERO IEEE_UNDERFLOW_FLAG

### DISCUSSION THREAD

**Discussion by walleludvig** [nrc-cnrc/EGSnrc#980] [Questions]
Hi,

I am seeking advice on how to deal with what seems to be a GNU compiler issue.

Every time I run a simulation with significant amount of histories (>1e6) the simulation breaks and reporting the following:

`Note: The following floating-point exceptions are signalling: IEEE_DIVIDE_BY_ZERO IEEE_UNDERFLOW_FLAG`

I have had a look around for records of similar errors and on the thread https://github.com/firemodels/fds/issues/5638 it is suggested to add a flag of
`-ffpe-trap=invalid,zero,overflow -ffpe-summary=none` at compilation. However, I am unsure how to implement this with an already configured EGSnrc system. 

Has anyone else ran into this issue with EGSnrc simulations and may know how to overcome it? I would hope to maybe add the flag mentioned above to compilation of one of my BEAMnrc accelerators but also not quite sure how to do exactly that.

Any help or advice is appreciated.

Thanks :D

**Comment by rtownson** [comment]
I haven't come across this before. You can set compiler flags in the `.conf` files in `HEN_HOUSE/specs/`, probably just to either the fortran or C++ flags depending on which application you're using. Then recompile the application (and `HEN_HOUSE/egs++` if you're using an egs++ app). Let us know if it helps.

**Comment by walleludvig** [comment]
Thanks for your response @rtownson. I attempted to add the flag mentioned above to my `osx.conf` file and a file named `Fortran-suse11.4.conf` and then recompile one of my BEAMnrc accelerators however no change was observed. Would you be able to please specify what application you suggest that I recompile exactly? (Apologies for my ignorance on the issue). I am trying to run BEAMnrc applications.

**Reply by rtownson** [reply]
Adding the flags to `FOPT =` in `osx.conf` should work, and then you just need to recompile the accelerator (`make clean; make` in the egs_home/BEAM_youaccel directory). You should see the new compiler flags in the output from the terminal. If you do see the new flags and it hasn't solved the problem, then we can look at other solutions.

Have you tried the example input files like `BEAM_EX16MVp`, do you have the same issue there?

**Reply by walleludvig** [reply]
Hi @rtownson, thanks very much for responding to my enquiry and apologies for taking so long to respond back. I will stay more active.

I added the flags to `FOPT` in `osx.conf` and I did see these listed in the output from the terminal, however it did not solve the problem. I tried with several combinations like; all flags, and just each of the 2 flags by themselves. I observed that:

- only `-ffpe-summary=none`: Simulation breaks similarly to before (after some number of histories) but just did not report on the problem (as the flag suggests).
- only `-ffpe-trab=invalid,zero,overflow`: Simulation breaks with the following error right after starting to read renormalised photoelectric cross sections. `Program received SIGKILL: Illegal instruction. Backtrace for this error: Program received signal SIGTRAP: Trace/breakpint trap.`
- both flags: Simulation breaks with the following error right after starting to read renormalised photoelectric cross sections. `Program received SIGKILL: Illegal instruction. Backtrace for this error: Program received signal SIGTRAP: Trace/breakpint trap.`

I removed these compiler flags again and tried with the example input file `BEAM_EX16MVp` as suggested and this simulation ran fine with large histories, suggesting there is probably an issue with my input file I suppose. Any chance you would be able to identify an issue with this input file without spending too much time on this? Here is my input file in `.txt` format:
[0.1265SSD_5pRe_100kV.txt](https://github.com/nrc-cnrc/EGSnrc/files/11413786/0.1265SSD_5pRe_100kV.txt)

**Reply by rtownson** [reply]
Can you also share your material file?

**Reply by walleludvig** [reply]
Apologies... Yes here it is:

[XSample.txt](https://github.com/nrc-cnrc/EGSnrc/files/11452796/XSample.txt)

**Reply by rtownson** [reply]
You could try switching to an older version of gcc. I didn't have any issues running 2e6 particles, on gcc 4.8.5 (very old, I know). Try gcc 8?

**Reply by walleludvig** [reply]
Hi @rtownson,

I will try this at a later point when I need to simulate an experiment with larger amount of histories locally. For now I will use a virtual machine with Linux to perform my larger simulations after having used my personal computer for troubleshooting the implementations of my simulations.

However thank you so much for your suggestions and responses to this thread. I will post an update once I have attempted to switch to an older version of gcc.

**Reply by walleludvig** [reply]
Hi @rtownson,

I am coming back to this issue as I still cannot simulate a large enough amount of histories for my purposes, however I am wondering if I quoted an incorrect upper range for number of histories earlier as I am also able to run the input file provided above with 2M histories, however anything significantly above that (e.g, 3M) crashes with the error 
`Note: The following floating-point exceptions are signalling: IEEE_DIVIDE_BY_ZERO`. I am wondering if you would be able to test it out again but n-histories bigger than before (e.g., 3M or 5M)?

I also wonder why you think an older gcc compiler might work better? I am currently using `gcc-13 (Homebrew GCC 13.1.0) 13.1.0`, and a bit hesitant to change this for my current installation of EGSnrc. I have a fair bit of work on my current install, that I would have to transfer and re-compile, and in the past I have struggled to make this work smoothly when re-installing EGSnrc to change compiler links. However, if need be I will definitely do it :)

**Comment by crcrewso** [comment]
Just taking a step back to basics, it's been a long time since I've looked at the gcc quirks on macs. Could you try recompiling with -mcmodel=large as an optimization argument and see what that changes? I did a lot of work to make sure that EGS could routinely handle 1e9 histories and I thought we had this licked

**Comment by walleludvig** [comment]
Hi @crcrewso,

Thanks for taking a look at this. I tried re-compiling my BEAMnrc accelerator after adding `-mcmodel=large` to `FOPT` in `HEN_HOUSE/specs/osx.conf`. The complete set of optimisation flags here was then `FOPT = -O2 -mtune=native -mcmodel=large`.

I now get a ton of errors when trying to compile my accelerator, all saying `error: invalid variant 'BLEAH'` occurring in different places. e.g., `error: invalid variant 'BLEAH' adrp    x23, lC9104@BLEAH`.

Does this tell you anything?

**Reply by crcrewso** [reply]
No, I haven't see that before, I was expecting you to rerun configure and add the flag everywhere you see -mcmodel=medium. But... I just did a quick google search. The large memory model might be broken on M1 macs

**Reply by walleludvig** [reply]
@crcrewso apologies for the confusion. Any alternative suggestions if this is broken for M1?

**Reply by crcrewso** [reply]
It's my confusion, not yours. 
I'm trying to get your input file working but I'm having some issues running pegsless with the material file you sent. Could you try uploading a working egsinp with the corresponding material file again? With my default Pegs4 file I was able to drive it up to 1e10 histories and it calculated quite quickly.

**Reply by walleludvig** [reply]
@crcrewso that's very interesting, thanks for trying that. I have recently run the input file in this zip with the corresponding pegs file successfully, hope this one works on your end too.
[inputfiles.zip](https://github.com/nrc-cnrc/EGSnrc/files/13505505/inputfiles.zip)

**Reply by crcrewso** [reply]
That would be my mistake then, for some reason I thought you were running pegsless. So that's entirely on me. I am now able to get your accelerator up and running with the provided input file, and can calculate well past 1e7 histories. 

Looking at the other conversation I see that you're using GCC 13, I'll have to do a bit of work installing it on my system as that branch is quite new.

**Reply by walleludvig** [reply]
Right okay, so if this works for you @crcrewso,  then it could likely be due to my GCC compilers, or perhaps OS? What OS are you running on?

**Reply by crcrewso** [reply]
I'm using Pop OS (basically Ubuntu 22.04 LTS) with gcc 12, I do have access to a couple of servers that I can install gcc 13 on for testing. 
I tend to gravitate to at least n-1 for gcc when it comes to production EGSnrc and other similar code bases. 

It's been a while since I played with a mac, does HomeBrew let you install multiple versions in parallel? 
If so you could configure EGS using gfortran-12 gcc-12 etc

**Reply by walleludvig** [reply]
@crcrewso That's helpful information. Not sure if it lets me have multiple versions installed, but worst case I'll just downgrade to GCC-12. I'll post here once tested. Thank you for your help so far.

**Reply by crcrewso** [reply]
After you switch to gcc-12, could you try adjusting HEN_HOUSE/omega/beamnrc/beam_nrc_user_macros.mortran 
edit the line 
` REPLACE {$MXSTACK} WITH {900000}   "maximum stack (much less if no brem split)`  by adding a few zeros 
then inside the BEAM model, run make clean, and then make again. When you run it again, at the top of the output you should see a "Max Stack:" number which should match this change.

**Reply by walleludvig** [reply]
Hi @crcrewso,

Apologies for the delayed response, I have been drenched in work lately and not had the opportunity to continue trying on this.

While I didn't downgrade my GCC compiler yet (will get onto that), I tried to change the line you mentioned, recompile my BEAM model however this only works when I add a single zero to that number making the line 
`REPLACE {$MXSTACK} WITH {9000000}   "maximum stack (much less if no brem split)`. When I add another zero then my BEAM model compiles successfully but when trying to run a simulation I get the error:

```
dyld[97782]: dyld cache '(null)' not loaded: syscall to map cache into shared region failed
dyld[97782]: Library not loaded: /usr/lib/libSystem.B.dylib
  Referenced from: <82AF55A7-207C-3C46-AB56-A9C8C0455986> /Users/ludvigwalle/EGSnrc/egs_home/bin/osx/BEAM_XSampleSpectra
  Reason: tried: '/Users/ludvigwalle/EGSnrc/HEN_HOUSE/egs++/dso/osx/libSystem.B.dylib' (no such file), '/libSystem.B.dylib' (no such file), '/usr/lib/libSystem.B.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/lib/libSystem.B.dylib' (no such file), '/usr/lib/libSystem.B.dylib' (no such file, no dyld cache), '/usr/local/lib/libSystem.B.dylib' (no such file)
zsh: abort      BEAM_XSampleSpectra -i 0.1265SSD_5pRe_120kV.egsinp -p XSample.pegs4dat
```

When I add any more than that my BEAM model does not compile at all and returns a ton of errors similar to this one at compilation
```
36259 |             wt(np) = 0
      |            1
Error: The function result on the lhs of the assignment at (1) must have the pointer attribute.
```

I doubt that this has to do with my GCC compiler, so was wondering if this information is of any help in trying determine the root of my simulation history issues?

**Reply by crcrewso** [reply]
I got looking at your simulation a bit more and am now wondering what your goal is. Using the same settings I was able to replicate your errors while simultaneously maintaining my ability to generate 1e8+ histories.

**Reply by walleludvig** [reply]
@crcrewso what do you mean by that? Which errors?

My goal for these simulations are to simulate my X-ray tube and collimator assembly. I am scoring a phase space file at the collimator exit. I am doing this for a range of kVp settings and then planning to use these phase space files as source inputs to other experimental simulations. I have been able to do this fine with 2e6 number of histories, however this only yields ~90k photons in the phase space file at max (120kVp) which is too few for my needs.

**Reply by crcrewso** [reply]
@walleludvig I was wondering, I noticed the 120 kVp source setting. That's a far lower incident dose than I remember anyone using EGSnrc for. If you're in the US penelope might be a better choice for you, if not geant4 might be. 

@ftessier @rtownson Do either of you have references for high accuracy kV energy papers that used EGSnrc? 


[Geant4](https://geant4.web.cern.ch/)
[PyPenelope](https://pypenelope.sourceforge.net/)
[FDA MCGPU (Penelope)](https://www.fda.gov/medical-devices/science-and-research-medical-devices/mcgpu-gpu-accelerated-monte-carlo-x-ray-imaging-simulator)

**Reply by walleludvig** [reply]
@crcrewso

Yes I am simulating X-ray spectra for a radiology setup, however given that BEAMnrc includes a geometry module specifically for the modelling of X-ray tubes I would expect that EGSnrc should be suitable. I have also found some papers aimed at benchmarking EGSnrc in low energy range for photon fluency/spectra, charged backscatter coefficients, mass energy-absorption coefficients & dosimetry. Couple of examples:
```
Ali, E. S. M., & Rogers, D. W. O. (2008). Benchmarking EGSnrc in the kilovoltage energy range against experimental measurements of charged particle backscatter coefficients. Physics in Medicine & Biology, 53(6), 1527.

Poludniowski, G., Omar, A., & Andreo, P. (2022). Calculating X-ray Tube Spectra: Analytical and Monte Carlo Approaches

Valdes-Cortez, C., Mansour, I., Rivard, M. J., Ballester, F., Mainegra-Hing, E., Thomson, R. M., & Vijande, J. (2021). A study of Type B uncertainties associated with the photoelectric effect in low-energy Monte Carlo simulations. Physics in Medicine & Biology, 66(10), 105014.
```

From my reading, EGSnrc should be okay for modelling the x-ray tube. Nonetheless, I am not sure how you were able to run my simulation with significant amount of histories but not me. I will attempt to downgrade by GCC compiler as previously suggested since the error thrown seems to be a compiler issue indeed. If you have any other potential suggestions I would love to hear about them :D

**Reply by crcrewso** [reply]
Looks like I have some weekend reading and testing in my future!

**Reply by rtownson** [reply]
Indeed, EGSnrc is accurate down to 1keV.

I have reproduced the same error using the input file provided, so it is a problem with the input file and not the compilers. Apologies for the very slow response to confirm that. 

I haven't isolated the issue yet, but I can run 5M histories with your input file if I turn off DBS.

**Reply by walleludvig** [reply]
@rtownson thank you very much for this information, that's very useful to know. Maybe there is something wrong with the way that I've set my DBS up? I will be testing some variations to this now.

**Reply by walleludvig** [reply]
@rtownson & @crcrewso, I think I have resolved my issues for running more than ~2M histories now. I am not entirely sure what the cause of the issue was, but I've got an input file that's working now. I realised that I had but my bremsstrahlung splitting rejection plane a centimetre closer than intended so made that change, but I do not understand that this difference would impact my ability to run some number of histories.

However, when I run `diff` on the 2 input files (one which can handle maximum ~1.5M and one which can seemingly handle any number of histories) I get the following differences:

```
4c4
< 2000000, 33, 97, 99.99, 2, 2000, 0, 0,  NCASE ETC.
---
> 10000000000, 33, 97, 99.99, 2, 2000, 0, 0,  NCASE ETC.
9c9
< 0, 0, 0.516, 0.005, 0, 0, 0,  0 , ECUT,PCUT,IREJCT,ESAVE
---
> 0, 0, 0.516, 0.005, 0, 0, ,  0 , ECUT,PCUT,IREJCT,ESAVE
103c103
<  Z(cm) from zero reference plane= 11.5
---
>  Z(cm) from zero reference plane= 12.5
```

Does this tell either of you about the root of the issue I was experiencing?

**Reply by rtownson** [reply]
@blakewalters any idea why this would be the case? The segmentation fault happens only with the rejection plane at some positions: works fine at 12.5, seg fault at 11.5 even with different seeds, hangs at 10.5, seg fault at 9.5..

**Reply by ojalaj** [reply]
Sorry to bump into this thread, but does it matter that in BEAMnrc manual p. 93 it says that "The rejection plane is meant to be placed in the air layer between the bottom of the last component in the accelerator geometry and the scoring plane."

In the provided .egsinp file and in your discussion above the plane in either within CM "filter" or in CM "collim".

**Reply by ojalaj** [reply]
@walleludvig did you give a try what I suggested above?
